File: DesignPatterns/CompositePattern/Company.java
import java.util.ArrayList; 
import java.util.List; 

interface Employee {
  public void showEmployeeDetails();
}

class Developer implements Employee {

  private String name;
  private long empId;
  private String position;

  public Developer(long empId, String name, String position) {
    this.empId = empId;
    this.name = name;
    this.position = position;
  }

  public void showEmployeeDetails() {
    System.out.println("Devloper: " + empId + "-" + name);
  }
}

class Manager implements Employee {
  private String name;
  private long empId;
  private String position;

  public Manager(long empId, String name, String position) {
    this.empId = empId;
    this.name = name;
    this.position = position;
  }

  public void showEmployeeDetails(){
    System.out.println("Manager : " + empId + "-" + name);
  }
}


class CompanyDirectory implements Employee {

  private List<Employee> empList = new ArrayList<Employee>();

  public void showEmployeeDetails() {
    for(Employee emp: empList){
      emp.showEmployeeDetails();
    }
  }

  public void addEmployee(Employee emp) {
    empList.add(emp);
  }

  public void removeEmployee(Employee emp) {
    empList.remove(emp);
  }
}

public class Company {
  public static void main(String[] args) {
    Developer d1 = new Developer(100, "Ravi", "Frontend developer");
    Developer d2 = new Developer(101, "Hari", "Backend developer");
    CompanyDirectory engCd  = new CompanyDirectory();
    engCd.addEmployee(d1);
    engCd.addEmployee(d2);

    Manager mg1 = new Manager(200, "Deepak", "Manager");
    Manager mg2 = new Manager(201, "Mani", "Senior Manager");
    CompanyDirectory accCd = new CompanyDirectory();
    accCd.addEmployee(mg1);
    accCd.addEmployee(mg2);

    CompanyDirectory cd = new CompanyDirectory();
    cd.addEmployee(engCd);
    cd.addEmployee(accCd);
    cd.showEmployeeDetails();

  }
}

File: DesignPatterns/Structural/Composite/Ruby/company.rb
class Employee
  attr_reader :children
  attr_accessor :parent

  def initialize
    @children = []
  end

  def add_employee(employee)
    @children.push(employee)
    employee.parent = self
  end

  def show_employee
    raise NotImplementedError
  end
end

class Manager < Employee
  attr_accessor :emp_id, :name, :position

  def initialize(emp_id, name, position)
    super()
    @emp_id = emp_id
    @name = name
    @position = position
  end

  def show_employee
    puts "#{self.emp_id} - #{self.name} - #{self.position}"
  end
end

class Developer < Employee

  attr_accessor :emp_id, :name, :position

  def initialize(emp_id, name, position)
    super()
    @emp_id = emp_id
    @name = name
    @position = position
  end

  def show_employee
    puts "#{self.emp_id} - #{self.name} - #{self.position}"
  end
end

class CompanyDirectory < Employee
  attr_accessor :name

  def initialize(name)
    super()
    @name = name
  end

  def show_employee
    puts "#{self.name}: "
    @children.each do |child|
      child.show_employee
    end
  end
end

d1 = Developer.new(100, "Ravi", "Frontend developer")
d2 = Developer.new(101, "Hari", "Backend developer")
eng_cd  = CompanyDirectory.new("Engineering")
eng_cd.add_employee(d1)
eng_cd.add_employee(d2)

mg1 = Manager.new(200, "Deepak", "Manager")
mg2 = Manager.new(201, "Mani", "Senior Manager")
acc_cd = CompanyDirectory.new("Account")
acc_cd.add_employee(mg1)
acc_cd.add_employee(mg2)

cd = CompanyDirectory.new("CompanyDirectory")
cd.add_employee(eng_cd)
cd.add_employee(acc_cd)
cd.show_employee
